pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionSnapshotCache.java

```diff
@@ -18,81 +18,377 @@
 
         // Prune the cache
-        while (snapshots.size() > maxSnapshotToCache) {
-            snapshots.pollFirstEntry();
+        if (numberOfSnapshots > maxSnapshotToCache) {
+            removeSingleEntryWithMinimumTotalDistanceToPreviousAndNext();
+        }
+    }
+
+    private void removeSingleEntryWithMinimumTotalDistanceToPreviousAndNext() {
+        updateSortedEntriesByTotalDistance();
+
+        SnapshotEntry minEntry = sortedSnapshots.first();
+        // Defensive check: minEntry should never be head or tail, as these are boundary entries that must be preserved
+        if (minEntry == head || minEntry == tail) {
+            throw new IllegalStateException("minEntry should not be head or tail boundary entry");
+        }
+        SnapshotEntry minEntryNext = minEntry.next;
+        SnapshotEntry minEntryPrevious = minEntry.prev;
+
+        // remove minEntry from the sorted set
+        sortedSnapshots.remove(minEntry);
+
+        // remove minEntryPrevious and minEntryNext from the sorted set since the distance will be updated
+        if (minEntryNext != tail) {
+            sortedSnapshots.remove(minEntryNext);
+        }
+        if (minEntryPrevious != head) {
+            sortedSnapshots.remove(minEntryPrevious);
+        }
+
+        // remove minEntry from the linked list
+        minEntryPrevious.setNext(minEntryNext);
+        minEntryNext.setPrev(minEntryPrevious);
+        numberOfSnapshots--;
+
+        // handle the case where the entry to remove is the last entry that has been sorted
+        if (lastSortedEntry == minEntry) {
+            lastSortedEntry = minEntryPrevious;
+        }
+
+        // update distanceToPrevious for the next entry
+        minEntryNext.setDistanceToPrevious(minEntryNext.distanceToPrevious + minEntry.distanceToPrevious);
+
+        // add entries back to the sorted set so that entries up to lastSortedEntry are sorted
+        if (minEntryNext != tail) {
+            sortedSnapshots.add(minEntryNext);
+        }
+        if (minEntryPrevious != head) {
+            sortedSnapshots.add(minEntryPrevious);
+        }
+    }
+
+    /**
+     * Maintains a sorted set of entries ordered by their total distance to adjacent entries.
+     * This method calculates the 'distanceToPrevious' field for both current and next entries before adding them to the
+     * sorted set. Subsequent calls to this method will continue processing from where the last entry was added.
+     */
+    private void updateSortedEntriesByTotalDistance() {
+        SnapshotEntry current = lastSortedEntry != null ? lastSortedEntry.next : head.next;
+        SnapshotEntry previousLoopEntry = null;
+        while (current != null) {
+            // calculate the distance to the previous snapshot entry
+            if (current.distanceToPrevious == -1) {
+                long distanceToPrevious =
+                        distanceFunction.applyAsLong(Range.open(current.prev.position, current.position));
```

Comment:
@lhotari Will there be a problem here?

Range.open(a,b) (open interval) is not additive in mathematics:

(a,b) + (b,c) != (a,c), missing the b position itself (strictly speaking, it's off-by-one / does not satisfy concatenation additivity).